File: kotlin/src/main/kotlin/com/bhengubv/circleai/companion/reasoning/InnerMonologue.kt
// InnerMonologue.kt
//
// Kotlin port of the CircleAI.Companion inner-monologue contract and its two
// concrete implementations — the C# reference is the exact spec:
//   IInnerMonologue            (HerJarvisContracts.cs, contract 13)
//   SelfReflection             (record)
//   TemplateInnerMonologue     (HerJarvisRealImplementations.cs, impl 13)
//   ReasoningLoopInnerMonologue(ReasoningLoopInnerMonologue.cs, Phase E1)
//
// Inner monologue = the companion's self-reflection over its current context.
// TemplateInnerMonologue is a deterministic narrative-template reflection with
// no model dependency; ReasoningLoopInnerMonologue drives a reasoning-capable
// LLM and captures its <think> trace as the thought.

package com.bhengubv.circleai.companion.reasoning

import com.bhengubv.circleai.inference.GenerationOptions
import com.bhengubv.circleai.inference.IChatGenerator
import com.bhengubv.circleai.models.ChatFragmentKind
import com.bhengubv.circleai.models.ChatMessage
import java.time.Instant
import java.util.UUID

// ---------------------------------------------------------------------------
// SelfReflection
// ---------------------------------------------------------------------------

/** A single reflective [thought] captured at [at]. Mirrors C# `SelfReflection`. */
data class SelfReflection(
    val thought: String,
    val at: Instant,
)

// ---------------------------------------------------------------------------
// IInnerMonologue
// ---------------------------------------------------------------------------

/** Self-reflection / inner monologue over a JSON context. */
interface IInnerMonologue {
    suspend fun reflectAsync(contextJson: String): SelfReflection
}

// ---------------------------------------------------------------------------
// TemplateInnerMonologue
// ---------------------------------------------------------------------------

/**
 * Narrative-template reflection over the context JSON. Picks one of three
 * frames deterministically from the context's hash, fills `{summary}` with the
 * first dozen words of the de-punctuated JSON and `{direction}` with a
 * keyword-inferred next step.
 *
 * Note on parity: the C# reference selects the frame from
 * `contextJson.GetHashCode()`, which is randomised per .NET process and so is
 * not reproducible across runs by design. We use Kotlin's stable
 * `String.hashCode()` (masked non-negative) so the selection is deterministic
 * and testable; the frame text, summary extraction, and direction inference
 * match the C# exactly.
 */
class TemplateInnerMonologue : IInnerMonologue {

    override suspend fun reflectAsync(contextJson: String): SelfReflection {
        val summary = summarise(contextJson)
        val direction = inferDirection(contextJson)
        val seed = contextJson.hashCode() and Int.MAX_VALUE
        val frame = FRAMES[seed % FRAMES.size]
        val thought = frame.replace("{summary}", summary).replace("{direction}", direction)
        return SelfReflection(thought, Instant.now())
    }

    private companion object {
        val FRAMES = arrayOf(
            "Observation: {summary}. Implication: this likely means {direction}.",
            "Looking at {summary}, the salient pattern is {direction}.",
            "Given {summary}, my next step is to {direction}.",
        )

        // Matches the C# regex character class [\{\}\[\]\"].
        val PUNCT = Regex("[\\{\\}\\[\\]\"]")

        fun summarise(json: String): String {
            val clean = PUNCT.replace(json, " ")
            val words = clean.split(' ').filter { it.isNotEmpty() }.take(12)
            return words.joinToString(" ")
        }

        fun inferDirection(json: String): String = when {
            json.contains("error", ignoreCase = true) -> "diagnose the failure first"
            json.contains("goal", ignoreCase = true) -> "advance toward the stated goal"
            json.contains("user", ignoreCase = true) -> "respond to the user"
            else -> "gather more context"
        }
    }
}

// ---------------------------------------------------------------------------
// ReasoningLoopInnerMonologue
// ---------------------------------------------------------------------------

/**
 * Inner monologue powered by a reasoning-capable LLM. Streams fragments from
 * [llm], routing REASONING fragments into the "thought" and CONTENT fragments
 * into the visible conclusion. Prefers the reasoning trace; falls back to the
 * visible content, then to "(no inner state)". Stream failures are swallowed so
 * a reflection is always produced.
 */
class ReasoningLoopInnerMonologue(private val llm: IChatGenerator) : IInnerMonologue {

    override suspend fun reflectAsync(contextJson: String): SelfReflection {
        val messages = listOf(
            ChatMessage(id = UUID.randomUUID().toString(), role = "system", content = REASONING_SYSTEM_PROMPT),
            ChatMessage(
                id = UUID.randomUUID().toString(),
                role = "user",
                content = "Context (raw JSON):\n$contextJson\n\nReflect on this in 2-3 sentences.",
            ),
        )
        val options = GenerationOptions(maxTokens = 256, temperature = 0.5f, includeReasoning = true)

        val reasoning = StringBuilder()
        val content = StringBuilder()
        try {
            llm.streamFragmentsAsync(messages, options).collect { frag ->
                if (frag.kind == ChatFragmentKind.REASONING) reasoning.append(frag.text)
                else content.append(frag.text)
            }
        } catch (ex: Exception) {
            // Match the C# reference: log and continue with whatever accumulated.
            System.err.println("[ReasoningLoopInnerMonologue] LLM stream failed: ${ex.message}")
        }

        // Prefer the reasoning trace as the "thought"; fall back to visible content.
        var thought = if (reasoning.isNotEmpty()) reasoning.toString().trim() else content.toString().trim()
        if (thought.isEmpty()) thought = "(no inner state)"
        return SelfReflection(thought, Instant.now())
    }

    private companion object {
        const val REASONING_SYSTEM_PROMPT =
            "You are this user's inner monologue. Reason carefully before responding. " +
                "Use <think>...</think> blocks for chain-of-thought. The visible answer " +
                "afterwards should be short and reflective — not a solution, an observation."
    }
}

File: swift/Sources/CircleAI/InnerMonologue.swift
// InnerMonologue.swift
//
// Port of CircleAI.Companion inner-monologue layer — the C# reference:
//   - IInnerMonologue + SelfReflection   (HerJarvisContracts.cs)
//   - TemplateInnerMonologue             (HerJarvisRealImplementations.cs)
//   - ReasoningLoopInnerMonologue        (ReasoningLoopInnerMonologue.cs)
//
// Contract #13: self-reflection / inner monologue.
// TemplateInnerMonologue fills a narrative template from a context summary and
// an inferred direction. ReasoningLoopInnerMonologue drives a reasoning-capable
// LLM (o1 / DeepSeek-R1 style) and captures the <think> stream as the thought.
//
// In-memory + deterministic (the template path). The reasoning path defers to
// the injected IChatGenerator.

import Foundation

// MARK: - SelfReflection

/// A single reflective thought stamped with the time it was produced.
public struct SelfReflection: Sendable, Equatable {
    public let thought: String
    public let at: Date

    public init(thought: String, at: Date) {
        self.thought = thought
        self.at = at
    }
}

// MARK: - IInnerMonologue

/// Contract #13 — self-reflection / inner monologue.
public protocol IInnerMonologue: AnyObject {
    /// Produce a reflective thought about the supplied context JSON.
    func reflect(contextJson: String) async throws -> SelfReflection
}

// MARK: - TemplateInnerMonologue

/// Narrative-template reflection over context. Summarises the raw JSON into a
/// short phrase, infers a "direction" from salient keywords, and folds both into
/// one of three template frames. Ported from `TemplateInnerMonologue`
/// (HerJarvisRealImplementations.cs).
///
/// Note on frame selection: the C# reference seeds frame choice with
/// `contextJson.GetHashCode()`, which .NET randomises per process — so the exact
/// frame is not reproducible across runs (in C# either). This port uses a
/// deterministic FNV-1a hash so the choice is stable for a given input; the
/// meaningful, testable behaviour (summary + direction substitution, and that
/// the output is one of the three frames) is preserved exactly.
public final class TemplateInnerMonologue: IInnerMonologue, @unchecked Sendable {
    static let frames: [String] = [
        "Observation: {summary}. Implication: this likely means {direction}.",
        "Looking at {summary}, the salient pattern is {direction}.",
        "Given {summary}, my next step is to {direction}.",
    ]

    public init() {}

    public func reflect(contextJson: String) async throws -> SelfReflection {
        let summary = Self.summarise(contextJson)
        let direction = Self.inferDirection(contextJson)
        let seed = Self.stableHash(contextJson) & Int(Int32.max)
        let frame = Self.frames[seed % Self.frames.count]
        let thought = frame
            .replacingOccurrences(of: "{summary}", with: summary)
            .replacingOccurrences(of: "{direction}", with: direction)
        return SelfReflection(thought: thought, at: Date())
    }

    /// Strips JSON punctuation, keeps the first 12 whitespace-separated words.
    static func summarise(_ json: String) -> String {
        var clean = json
        for ch in ["{", "}", "[", "]", "\""] {
            clean = clean.replacingOccurrences(of: ch, with: " ")
        }
        let words = clean.split(separator: " ", omittingEmptySubsequences: true).prefix(12)
        return words.joined(separator: " ")
    }

    /// Salient-keyword direction, checked in the reference's order:
    /// error → goal → user → default.
    static func inferDirection(_ json: String) -> String {
        if json.range(of: "error", options: .caseInsensitive) != nil { return "diagnose the failure first" }
        if json.range(of: "goal", options: .caseInsensitive) != nil { return "advance toward the stated goal" }
        if json.range(of: "user", options: .caseInsensitive) != nil { return "respond to the user" }
        return "gather more context"
    }

    /// Deterministic 32-bit FNV-1a hash over the UTF-8 bytes, returned as a
    /// non-negative Int. Replaces .NET's per-process-randomised String.GetHashCode
    /// with a stable value so frame selection is reproducible.
    static func stableHash(_ s: String) -> Int {
        var hash: UInt32 = 2166136261
        for byte in s.utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 16777619
        }
        return Int(hash & 0x7FFF_FFFF)
    }
}

// MARK: - ReasoningLoopInnerMonologue

/// Inner-monologue powered by a reasoning-capable LLM. Streams fragments from
/// the injected `IChatGenerator`, accumulating the `.reasoning` fragments as the
/// inner thought and `.content` fragments as the visible conclusion; prefers the
/// reasoning trace, falls back to content, and degrades to "(no inner state)".
/// Ported from `ReasoningLoopInnerMonologue` (ReasoningLoopInnerMonologue.cs).
public final class ReasoningLoopInnerMonologue: IInnerMonologue, @unchecked Sendable {
    private static let reasoningSystemPrompt =
        "You are this user's inner monologue. Reason carefully before responding. " +
        "Use <think>...</think> blocks for chain-of-thought. The visible answer " +
        "afterwards should be short and reflective — not a solution, an observation."

    private let llm: IChatGenerator

    public init(_ llm: IChatGenerator) {
        self.llm = llm
    }

    public func reflect(contextJson: String) async throws -> SelfReflection {
        let messages = [
            ChatMessage(role: "system", content: Self.reasoningSystemPrompt),
            ChatMessage(role: "user",
                        content: "Context (raw JSON):\n\(contextJson)\n\nReflect on this in 2-3 sentences."),
        ]
        let options = GenerationOptions(maxTokens: 256, temperature: 0.5, includeReasoning: true)

        var reasoning = ""
        var content = ""
        // The reference wraps the stream in try/catch and swallows failures,
        // degrading to whatever was captured (or the placeholder below).
        for await frag in llm.streamFragments(messages: messages, options: options) {
            if frag.kind == .reasoning {
                reasoning += frag.text
            } else {
                content += frag.text
            }
        }

        // Prefer the reasoning trace as the "thought"; fall back to visible content.
        var thought = !reasoning.isEmpty
            ? reasoning.trimmingCharacters(in: .whitespacesAndNewlines)
            : content.trimmingCharacters(in: .whitespacesAndNewlines)
        if thought.isEmpty { thought = "(no inner state)" }
        return SelfReflection(thought: thought, at: Date())
    }
}
